source.go

```diff
@@ -303,12 +338,28 @@ func WithCreateDestPath() llb.CopyOption {
 	})
 }
 
-func SourceIsDir(src Source) bool {
+func (s *Source) IsDir(srcState llb.State, sOpt SourceOpts) bool {
+	return SourceIsDir(*s, srcState, sOpt)
+}
+
+func SourceIsDir(src Source, srcState llb.State, sOpt SourceOpts) bool {
 	switch {
+	case src.Context != nil:
+		if !isRoot(src.Path) {
```

Comment:
All this logic needs to apply for *all* source types (which are normally considered a dir).